internals-js/src/__tests__/extractSubgraphsFromSupergraph.test.ts

```diff
@@ -597,3 +597,86 @@ test('throw meaningful error for invalid federation directive fieldSet', () => {
     + '[serviceB] On field "A.a", for @requires(fields: "b { x }"): Cannot query field "b" on type "A" (if the field is defined in another subgraph, you need to add it to this subgraph with @external).'
   );
 })
+
+
+test('throw meaningful error for type erased from supergraph due to extending an entity without a key', () => {
+  // Supergraph generated by fed1 composition from:
+  // ServiceA:
+  //     type Query {
+  //       t: T
+  //     }
+  //
+  //     type T @key(fields: "id") {
+  //       id: ID!
+  //       x: Int!
+  //     }
+  //
+  // ServiceB
+  //     extend type T {
+  //       x: Int! @external
+  //     }
+  //
+  //     type Other {
+  //       f: T @provides(fields: "x")
+  //     }
+  //
+  // The issue of that schema is that `T` is referenced in `ServiceB`, but because it extends an entity type
+  // without a key and has noly external field, ther is no remaining traces of it's definition in `ServiceB`
+  // in the supergraph. As extract cannot make up the original definition out of thin air, it ends up erroring
+  // when extracting `Other.t` due to not knowing that type.
+  const supergraph = `
+    schema
+      @core(feature: "https://specs.apollo.dev/core/v0.2"),
+      @core(feature: "https://specs.apollo.dev/join/v0.1", for: EXECUTION)
+    {
+      query: Query
+    }
+
+    directive @core(as: String, feature: String!, for: core__Purpose) repeatable on SCHEMA
+
+    directive @join__field(graph: join__Graph, provides: join__FieldSet, requires: join__FieldSet) on FIELD_DEFINITION
+
+    directive @join__graph(name: String!, url: String!) on ENUM_VALUE
+
+    directive @join__owner(graph: join__Graph!) on INTERFACE | OBJECT
+
+    directive @join__type(graph: join__Graph!, key: join__FieldSet) repeatable on INTERFACE | OBJECT
+
+    type Other {
+      f: T @join__field(graph: SERVICEB, provides: "x")
+    }
+
+    type Query {
+      t: T @join__field(graph: SERVICEA)
+    }
+
+    type T
+      @join__owner(graph: SERVICEA)
+      @join__type(graph: SERVICEA, key: "id")
+    {
+      id: ID! @join__field(graph: SERVICEA)
+      x: Int! @join__field(graph: SERVICEA)
+    }
+
+    enum core__Purpose {
+      EXECUTION
+      SECURITY
+    }
+
+    scalar join__FieldSet
+
+    enum join__Graph {
+      SERVICEA @join__graph(name: "serviceA" url: "")
+      SERVICEB @join__graph(name: "serviceB" url: "")
+    }
+  `;
+
+  const schema = buildSupergraphSchema(supergraph)[0];
+  expect(() => extractSubgraphsFromSupergraph(schema)).toThrow(
+    'Error extracting subgraphs from the supergraph: this might be due to errors in subgraphs that were mistakenly ignored by federation 0.x versions but are rejected by federation 2.\n'
+    + 'Please try composing your subgraphs with federation 2: this should help precisely pinpoint the problems and, once fixed, generate a correct federation 2 supergraph.\n'
```

Comment:
**observation** I wonder if we'll eventually be able to make this a suggestion the user will be able to act on (i.e. `Add a @link to the federation 2 schema`, etc). This message is definitely a good direction, but I might be a little lost as a user. Maybe if a tool like rover sees this message, it could try to recompose with fed2 automagically.

Anyway, nothing to do here but thought I'd mention it while I was thinking about it.

Reply:
The thing is that user are going to run into this error mostly only when they run a gateway/router with a fed1 supergraph. So `rover` is probably not in the picture here (that is, `rover` does composition, but if you compose with the fed 2 code base, you never get a fed1 supergraph and so that doesn't apply), and the `gateway`/`router` usually don't really have any access to the subgraphs, so I'm not sure how much more can be done.

I mean, I'm obviously open to further improvements, it's just not obvious what more we can do (well, unless we can to consider pretty complex stuff like having the `gateway`/`router` reach to studio somehow and do some magic, but I'm not even sure that would desirable).

> i.e. `Add a @link to the federation 2 schema`

I'll note that this isn't the issue here. It's perfectly for the user running into this to keep its schema as a fed1 schema initially; the problem is that there is bugs/dodgy stuff in their fed1 schema which fed 1 didn't detect and that's what they should fix. In fact, my advise would be that they _first_ fix the problem in the fed1 schema and only _then_ consider migrating the schema into a fed 2 schema.

Reply:
Got it, that makes sense. I appreciate the explanation.